actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,54 @@
+module ActionView
+  module CollectionCaching
+    extend ActiveSupport::Concern
+
+    included do
+      # Overriden in Railtie then set to Rails.cache which won't be defined now.
+      mattr_reader(:collection_cache) { ActiveSupport::Cache::NullStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }
+      rendered_partials = @collection.any? ? yield : []
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      ActionController::Base.perform_caching && (@options[:cache].present? || @locals[:cache].present?)
+    end
+
+    def collection_by_cache_keys
+      seed = @options[:cache].respond_to?(:call) ? @options[:cache] : ->(i) { i }
+
+      Hash[ @collection.map { |item| [expanded_cache_key(seed.call(item)), item] } ]
+    end
+
+    def expanded_cache_key(key)
+      caching_controller.fragment_cache_key(@view.cache_fragment_name(key)).tap do |k|
+        return k.dup if k.frozen? # #read_multi & #write may require mutability, Dalli 2.6.0.
```

Comment:
Agree that doing an explicit `return` within a `#tap` is confusing, since tap is designed to do just the opposite: return the receiver. Clearer to take its return value and do an explicit conditional: `v = …; v.frozen? ? v.dup : v`